zapcore/sampler.go

```diff
@@ -126,9 +178,35 @@ func (s *sampler) Check(ent Entry, ce *CheckedEntry) *CheckedEntry {
 	}
 
 	counter := s.counts.get(ent.Level, ent.Message)
-	n := counter.IncCheckReset(ent.Time, s.tick)
+	n, prevN := counter.IncCheckReset(ent.Time, s.tick)
 	if n > s.first && (n-s.first)%s.thereafter != 0 {
 		return ce
 	}
+	if s.reporting.enabled && prevN > 0 {
+		s.report(ent, prevN)
```

Comment:
`With` seems to be the wrong time to be doing this reporting, better if it could be right at checked entry write, which would then support proper error propagation (as noted below).

Reply:
Again, this is `Check()` not `With()`, and this is where we check counters value. So it got to be here.